File: mobile/welcome-screen/android-preview-project/app/src/main/java/com/showup/welcome/CountryPicker.kt
/*
 * CountryPicker.kt
 * ShowUp · the flag mark and the country list sheet (SHOWUP-143)
 *
 * The ticket puts the country list itself out of scope, so this is the smallest thing that lets a
 * user actually change their dial code: a bottom sheet, one row per country, the current one
 * marked. No search — at this length the list scrolls fine, and a search field is a second input to
 * design and test for no benefit yet.
 */
package com.showup.welcome

import androidx.compose.foundation.Canvas
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.drawscope.Stroke

import com.showup.designsystem.PrimaryButton
import com.showup.designsystem.PrimaryButtonVariant

import com.showup.designsystem.Spacing

import android.graphics.BitmapFactory
import androidx.compose.foundation.Image
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.showup.designsystem.Border
import com.showup.designsystem.Elevated
import com.showup.designsystem.Fg
import com.showup.designsystem.Manrope
import com.showup.designsystem.Muted
import com.showup.designsystem.Orange
import com.showup.designsystem.Purple
import com.showup.designsystem.Raised

/**
 * The flag for a country, from the bundled artwork.
 *
 * Drawn flags were replaced by real artwork once the list went from 35 countries to every country.
 * Hand-drawing did not scale and was not honest about it: nine were drawn by hand and two came out
 * wrong — Canada as a spiky asterisk, Portugal as a logo. At 250 that rate means dozens wrong, and
 * a wrong national flag is not a cosmetic bug.
 *
 * These are the flag-icons set (MIT, see assets/flags/LICENSE-flag-icons.txt), rasterised to 96x72
 * — enough for a 22dp mark at four times density. 257 files, 479 KB in total, about 1.9 KB each.
 *
 * A country with no artwork falls back to its ISO code rather than an empty box, so a gap looks
 * deliberate instead of broken.
 */
@Composable
fun Flag(country: Country, width: Dp = 22.dp) {
    val height = width * 14f / 22f
    val shape = RoundedCornerShape(2.dp)
    val context = LocalContext.current

    // Decoded once per country and kept: the picker scrolls through hundreds of rows, and decoding
    // a bitmap on every frame of a fling is exactly how a list starts to stutter.
    val bitmap = remember(country.iso) { loadFlag(context, country.iso) }

    val box = Modifier
        .size(width = width, height = height)
        .clip(shape)
        .border(0.5.dp, Color(0x591D1129), shape)

    if (bitmap != null) {
        Image(
            bitmap = bitmap,
            // Decorative: the country's name sits next to it in the list, and the dial code is on
            // the pill, so a screen reader gains nothing from "flag of Germany".
            contentDescription = null,
            modifier = box,
            contentScale = ContentScale.Crop,
        )
    } else {
        Box(box.background(Raised), contentAlignment = Alignment.Center) {
            Text(
                country.iso, color = Muted, fontFamily = Manrope,
                fontWeight = FontWeight.Bold, fontSize = 8.sp, letterSpacing = 0.5.sp,
            )
        }
    }
}

/** Null when the country has no bundled flag, which the caller renders as the ISO code. */
private fun loadFlag(context: android.content.Context, iso: String): ImageBitmap? = try {
    context.assets.open("flags/" + iso.uppercase() + ".png").use { stream ->
        BitmapFactory.decodeStream(stream)?.asImageBitmap()
    }
} catch (e: java.io.IOException) {
    null
}

/**
 * The list, as a bottom sheet over the screen.
 *
 * Opens scrolled to the current country rather than at the top — the user is here to change a
 * value they can already see, so showing them where they are is the first useful thing.
 */
@Composable
fun CountrySheet(
    current: Country,
    onPick: (Country) -> Unit,
    onDismiss: () -> Unit,
) {
    val listState = rememberLazyListState()
    LaunchedEffect(Unit) {
        val i = COUNTRIES.indexOfFirst { it.iso == current.iso }
        if (i > 0) listState.scrollToItem(i)
    }

    Box(Modifier.fillMaxSize()) {
        // Tappable here, unlike the conflict modal on 144: this sheet is a convenience with a
        // harmless dismiss, not a decision the user has to make.
        Box(
            Modifier
                .fillMaxSize()
                .background(Color(0x6B140C1E))
                .clickable(onClick = onDismiss)
        )
        Column(
            Modifier
                .align(Alignment.BottomCenter)
                .fillMaxWidth()
                .fillMaxHeight(0.72f)
                .clip(RoundedCornerShape(topStart = 28.dp, topEnd = 28.dp))
                .background(Elevated),
        ) {
            Box(Modifier.fillMaxWidth().padding(top = Spacing.lg), contentAlignment = Alignment.Center) {
                Box(Modifier.size(width = 40.dp, height = 4.dp)
                    .clip(RoundedCornerShape(50)).background(Border))
            }
            Text(
                "Choose your country",
                Modifier.padding(start = Spacing.screenGutter, end = Spacing.screenGutter, top = Spacing.xxl, bottom = Spacing.xl),
                color = Fg, fontFamily = Manrope, fontWeight = FontWeight.Bold, fontSize = 17.sp,
            )
            LazyColumn(state = listState, modifier = Modifier.weight(1f)) {
                items(COUNTRIES, key = { it.iso }) { c ->
                    val selected = c.iso == current.iso
                    Row(
                        Modifier
                            .fillMaxWidth()
                            .clickable { onPick(c) }
                            .background(if (selected) Orange.copy(alpha = 0.07f) else Color.Transparent)
                            .padding(horizontal = Spacing.screenGutter, vertical = 13.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(14.dp),
                    ) {
                        Flag(c)
                        Text(
                            c.name, Modifier.weight(1f),
                            color = Fg, fontFamily = Manrope,
                            fontWeight = if (selected) FontWeight.Bold else FontWeight.Medium,
                            fontSize = 15.sp,
                        )
                        Text(
                            c.dial,
                            color = if (selected) Purple else Muted, fontFamily = Manrope,
                            fontWeight = FontWeight.SemiBold, fontSize = 15.sp,
                        )
                        if (selected) Icon(BrandIcon.Check, 16.dp, tint = Purple, strokeWidth = 2.4.dp)
                        else Spacer(Modifier.width(Spacing.xxl))
                    }
                }
            }
            // Clear of the gesture bar without the sheet needing to know about insets.
            Box(
                Modifier
                    .fillMaxWidth()
                    .windowInsetsPadding(WindowInsets.safeDrawing)
                    .padding(horizontal = Spacing.screenGutter, vertical = Spacing.xl)
            ) {
                PrimaryButton("Close", onDismiss, variant = PrimaryButtonVariant.Ghost)
            }
        }
    }
}

File: mobile/welcome-screen/ios-app/ShowUpWelcome/CountryPicker.swift
//  CountryPicker.swift
//  ShowUp · the flag mark and the country list sheet (SHOWUP-143)
//
//  The ticket puts the country list itself out of scope, so this is the smallest thing that lets a
//  user actually change their dial code: a sheet, one row per country, the current one marked. No
//  search — at this length the list scrolls fine, and a search field is a second input to design
//  and test for no benefit yet.

import SwiftUI

/// The flag for a country, from the bundled artwork.
///
/// Drawn flags were replaced by real artwork once the list went from 35 countries to every country.
/// Hand-drawing did not scale and was not honest about it: nine were drawn by hand and two came out
/// wrong — Canada as a spiky asterisk, Portugal as a logo. At 250 that rate means dozens wrong, and
/// a wrong national flag is not a cosmetic bug.
///
/// These are the flag-icons set (MIT, see Flags/LICENSE-flag-icons.txt), rasterised to 96x72 —
/// enough for a 22pt mark at three times density. 257 files, 479 KB in total.
///
/// A country with no artwork falls back to its ISO code rather than an empty box, so a gap looks
/// deliberate instead of broken.
struct FlagView: View {
    let country: Country
    var width: CGFloat = 22

    private var height: CGFloat { width * 14 / 22 }

    var body: some View {
        Group {
            if let image = FlagView.image(country.iso) {
                Image(uiImage: image)
                    .resizable()
                    // Decorative: the country's name sits next to it in the list and the dial code
                    // is on the pill, so a screen reader gains nothing from "flag of Germany".
                    .accessibilityHidden(true)
                    .scaledToFill()
            } else {
                ZStack {
                    Color.liqRaised
                    Text(country.iso)
                        .font(F.manrope(8, .bold))
                        .tracking(0.5)
                        .foregroundColor(.liqMuted)
                }
            }
        }
        .frame(width: width, height: height)
        .clipShape(RoundedRectangle(cornerRadius: 2, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: 2, style: .continuous)
                .strokeBorder(Color.liqFg.opacity(0.35), lineWidth: 0.5)
        )
    }

    /// Decoded once per country and kept. The picker scrolls through hundreds of rows, and decoding
    /// a bitmap on every frame of a fling is exactly how a list starts to stutter.
    ///
    /// @MainActor because Swift 6 rejects nonisolated global mutable state outright, and a shared
    /// cache is exactly that. Safe to isolate rather than rework: the only thing that ever touches
    /// it is FlagView's own body, and a SwiftUI View is already on the main actor. So this is an
    /// annotation that tells the compiler where the code already runs -- no behaviour changes and
    /// no call site moves.
    @MainActor private static var cache: [String: UIImage] = [:]

    @MainActor static func image(_ iso: String) -> UIImage? {
        let key = iso.uppercased()
        if let hit = cache[key] { return hit }
        guard let url = Bundle.main.url(forResource: key, withExtension: "png", subdirectory: "Flags"),
              let data = try? Data(contentsOf: url),
              let image = UIImage(data: data)
        else { return nil }
        cache[key] = image
        return image
    }
}

/// The list, as a sheet over the screen.
///
/// Opens scrolled to the current country rather than at the top — the user is here to change a
/// value they can already see, so showing them where they are is the first useful thing.
struct CountrySheet: View {
    let current: Country
    let onPick: (Country) -> Void
    let onDismiss: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            Capsule()
                .fill(Color.liqBorder)
                .frame(width: 40, height: 4)
                .padding(.top, Spacing.lg)

            Text("Choose your country")
                .font(F.manrope(17, .bold))
                .foregroundColor(.liqFg)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, Spacing.screenGutter)
                .padding(.top, Spacing.xxl)
                .padding(.bottom, Spacing.xl)

            ScrollViewReader { proxy in
                ScrollView {
                    LazyVStack(spacing: 0) {
                        ForEach(COUNTRIES) { c in
                            let selected = c.iso == current.iso
                            Button {
                                onPick(c)
                            } label: {
                                HStack(spacing: 14) {
                                    FlagView(country: c)
                                    Text(c.name)
                                        .font(F.manrope(15, selected ? .bold : .medium))
                                        .foregroundColor(.liqFg)
                                        .frame(maxWidth: .infinity, alignment: .leading)
                                    Text(c.dial)
                                        .font(F.manrope(15, .semibold))
                                        .foregroundColor(selected ? .liqPurple : .liqMuted)
                                    if selected {
                                        BrandIconView(icon: .check, size: 16, stroke: 2.4, tint: .liqPurple)
                                    } else {
                                        Color.clear.frame(width: 16, height: 16)
                                    }
                                }
                                .padding(.horizontal, Spacing.screenGutter)
                                .padding(.vertical, 13)
                                .background(selected ? Color.liqOrange.opacity(0.07) : Color.clear)
                                .contentShape(Rectangle())
                            }
                            .buttonStyle(.plain)
                            .id(c.iso)
                        }
                    }
                }
                .onAppear { proxy.scrollTo(current.iso, anchor: .center) }
            }

            PrimaryButton("Close", variant: .ghost, action: onDismiss)
                .padding(.horizontal, Spacing.screenGutter)
                .padding(.vertical, Spacing.xl)
        }
        .background(Color.liqElevated)
    }
}
